{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6432e006",
   "metadata": {},
   "source": [
    "## Purpose of the Analysis\n",
    "\n",
    "1. Identify proteins meeting the following criteria\n",
    "\n",
    " - negatively affect dengue virus replication when inhibited by siRNA*\n",
    " AND\n",
    " - are either:\n",
    "        - modulated in protein abundance\n",
    "        - modulated in rna abundance\n",
    "        - interact with the dengue virus\n",
    "        \n",
    "2. Identify systems of interacting proteins enriched in those that affect dengue replication\n",
    "\n",
    "* \\*the highest the Zscore, the more these factors negatively affect dengue virus replication\n",
    "\n",
    "3. Characterize the systems:\n",
    " - Name them by top hits from gProfiler enrichment\n",
    " - Name them using GPT-4\n",
    " - Analyze the systems with GPT-4\n",
    "     - General analysis\n",
    "     - Specific analyses in which data associated with the proteins is provided. This will be a matter of experimentation. Probably most tractable for small systems. We can try specifically asking for interpretation of the system for its role in response to dengue infection.\n",
    "\n",
    "4. Other points:\n",
    " - How should we interpret the 24 vs 48 hour data?\n",
    " - What visualization strategies will work best?\n",
    "\n",
    "\n",
    "## Data Loading\n",
    " - Used ChatGPT data analyis to merge the sheets in the original excel file provided by Laura\n",
    "     - Added a boolean column \"binds_viral_protein\" based on the viral interaction data\n",
    "     - Added boolean columns indicating whether the protein had significant values for the various measurements\n",
    "\n",
    "## Hierarchical Model Building\n",
    " - Made an network of interactions between the proteins: STRING HC\n",
    "     - Next step - add dengue AP-MS\n",
    " - Made a hierarchical from the subnetwork interconnecting the proteins of interest\n",
    " - Named the model systems by enrichment\n",
    " \n",
    "## TODO\n",
    " - Name and some or all of the model systems by LLM\n",
    " - Identify goal proteins\n",
    " - Identify systems enriched with goal proteins\n",
    " - Style the model and the interactome to highlight the data\n",
    " - Generate reports\n",
    " - Generate figures\n",
    " \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe2451b",
   "metadata": {},
   "source": [
    "## Experimental Protocol\n",
    "We took primary human dendritic cells, infected them with dengue virus (serotype 3) and subjected these cells to either: \n",
    " - siRNA screening to identify human host factors that act to restrict viral replication, \n",
    " - Proteomics (Protein Abundance) to look at human proteins that change in abundance following infection. \n",
    "   - This was done at 24h and 48h post infection. (Jeff Johnson, Krogan lab)\n",
    " - RNAseq to look at cellular mRNAs that are differently expressed following infection.\n",
    "   - This was done at 24h and 48h post infection. (Stephen Wolinski, NWU)\n",
    "   \n",
    "We also include host-dengue PPI data from (Shah et al., Cell 2018). \n",
    "Priya Shah was at the Krogan lab when this data was collected\n",
    "\n",
    "The datasets are already thresholded based maximum overlap and significance, and all have GeneID as common identifier. \n",
    "\n",
    "20220202_Thresholded_Datasets_DHIPC.xlsx\n",
    "\n",
    "Shah et al., Dengue Interactions\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10664e96",
   "metadata": {},
   "source": [
    "## Get the UniProt and HGNC IDs based on the NCBI Gene IDs\n",
    "\n",
    "Load the unified dengue spreadsheet (Created using ChatGPT data analysis mode)\n",
    "\n",
    "Add two columns, one for UniProt and one for HGNC.\n",
    "\n",
    "Noted gene ids not found by MyGene.info. Not sure what the story is for 441956 and 45338"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3add040b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3437\n",
      "{'query': '441956', 'notfound': True}\n",
      "gene id 441956 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "{'query': '45338', 'notfound': True}\n",
      "gene id 45338 was not found\n",
      "IFIT3\n",
      "0         IFIT3\n",
      "1        UBE2L6\n",
      "2        EIF4G3\n",
      "3        MRPS14\n",
      "4         UBXN7\n",
      "         ...   \n",
      "2045     ZNF594\n",
      "2046     ZNF623\n",
      "2047    ZNF702P\n",
      "2048     ZNF711\n",
      "2049     ZNF724\n",
      "Name: GeneID, Length: 2050, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>siRNA_GeneSymbol</th>\n",
       "      <th>DV3_24h-Mock_24h</th>\n",
       "      <th>DV3_48h-Mock_48h</th>\n",
       "      <th>siRNA_Screen_Average_Zscore</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>Condition</th>\n",
       "      <th>GeneSymbol_48hpi</th>\n",
       "      <th>...</th>\n",
       "      <th>dengue_MiST_list</th>\n",
       "      <th>PPI_GeneSymbol</th>\n",
       "      <th>viral_interaction</th>\n",
       "      <th>has_siRNA</th>\n",
       "      <th>has_protein_24hr</th>\n",
       "      <th>has_protein_48hr</th>\n",
       "      <th>has_rnaSeq_24hr</th>\n",
       "      <th>has_rnaSeq_48hr</th>\n",
       "      <th>UniProtID</th>\n",
       "      <th>HGNC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3437</td>\n",
       "      <td>O14879</td>\n",
       "      <td>IFIT3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.494644</td>\n",
       "      <td>5.698257</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.197744</td>\n",
       "      <td>24h</td>\n",
       "      <td>IFIT3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>O14879</td>\n",
       "      <td>IFIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9246</td>\n",
       "      <td>O14933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.504380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>O14933</td>\n",
       "      <td>UBE2L6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8672</td>\n",
       "      <td>O43432</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.897606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>O43432</td>\n",
       "      <td>EIF4G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>63931</td>\n",
       "      <td>O60783</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.707202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>O60783</td>\n",
       "      <td>MRPS14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26043</td>\n",
       "      <td>O94888</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.721373</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>O94888</td>\n",
       "      <td>UBXN7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneID UniprotID GeneSymbol siRNA_GeneSymbol  DV3_24h-Mock_24h  \\\n",
       "0    3437    O14879      IFIT3              NaN          3.494644   \n",
       "1    9246    O14933        NaN              NaN               NaN   \n",
       "2    8672    O43432        NaN              NaN               NaN   \n",
       "3   63931    O60783        NaN              NaN               NaN   \n",
       "4   26043    O94888        NaN              NaN               NaN   \n",
       "\n",
       "   DV3_48h-Mock_48h  siRNA_Screen_Average_Zscore    log2FC Condition  \\\n",
       "0          5.698257                          NaN  8.197744       24h   \n",
       "1          2.504380                          NaN       NaN       NaN   \n",
       "2         -1.897606                          NaN       NaN       NaN   \n",
       "3          0.707202                          NaN       NaN       NaN   \n",
       "4         -2.721373                          NaN       NaN       NaN   \n",
       "\n",
       "  GeneSymbol_48hpi  ...  dengue_MiST_list PPI_GeneSymbol viral_interaction  \\\n",
       "0            IFIT3  ...               NaN            NaN                 0   \n",
       "1              NaN  ...               NaN            NaN                 0   \n",
       "2              NaN  ...               NaN            NaN                 0   \n",
       "3              NaN  ...               NaN            NaN                 0   \n",
       "4              NaN  ...               NaN            NaN                 0   \n",
       "\n",
       "  has_siRNA has_protein_24hr  has_protein_48hr  has_rnaSeq_24hr  \\\n",
       "0         0                1                 1                1   \n",
       "1         0                0                 1                0   \n",
       "2         0                0                 1                0   \n",
       "3         0                0                 1                0   \n",
       "4         0                0                 1                0   \n",
       "\n",
       "   has_rnaSeq_48hr  UniProtID    HGNC  \n",
       "0                1     O14879   IFIT3  \n",
       "1                0     O14933  UBE2L6  \n",
       "2                0     O43432  EIF4G3  \n",
       "3                0     O60783  MRPS14  \n",
       "4                0     O94888   UBXN7  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "# Fetch the UniProt IDs and HGNC symbols for a list of NCBI GeneIDs using the MyGene.info API\n",
    "def fetch_uniprot_ids_and_hgnc_via_mygene(gene_ids):\n",
    "    url = \"https://mygene.info/v3/query\"\n",
    "    headers = {\"content-type\": \"application/x-www-form-urlencoded\"}\n",
    "    data = {\n",
    "        \"q\": \",\".join(map(str, gene_ids)),\n",
    "        \"species\": \"human\",\n",
    "        \"fields\": \"uniprot,symbol\"\n",
    "    }\n",
    "    \n",
    "    response = requests.post(url, headers=headers, data=data)\n",
    "    results = response.json()\n",
    "\n",
    "    # Extract UniProt IDs and HGNC symbols from the results\n",
    "    uniprot_ids = {}\n",
    "    hgnc_symbols = {}\n",
    "    for hit in results:\n",
    "        if \"_id\" in hit:\n",
    "            #print(hit)\n",
    "            gene_id = hit[\"_id\"]\n",
    "            uniprot_id = hit.get(\"uniprot\", {}).get(\"Swiss-Prot\")\n",
    "            hgnc_symbol = hit.get(\"symbol\")\n",
    "            uniprot_ids[gene_id] = uniprot_id\n",
    "            hgnc_symbols[gene_id] = hgnc_symbol\n",
    "        else:\n",
    "            print(hit)\n",
    "            print(f'gene id {hit.get(\"query\")} was not found')\n",
    "\n",
    "    return uniprot_ids, hgnc_symbols\n",
    "\n",
    "# Read the input data (replace with your file path)\n",
    "df = pd.read_excel(\"dengue_09162023.xlsx\", sheet_name=\"Sheet1\")\n",
    "\n",
    "# Fetch UniProt IDs and HGNC symbols for all GeneIDs using MyGene.info in batches\n",
    "gene_ids_df = df[\"GeneID\"].astype(str)\n",
    "gene_ids = gene_ids_df.tolist()\n",
    "print(gene_ids[0])\n",
    "uniprot_ids, hgnc_symbols = fetch_uniprot_ids_and_hgnc_via_mygene(gene_ids)\n",
    "print(hgnc_symbols.get(\"3437\"))\n",
    "print(gene_ids_df.map(hgnc_symbols))\n",
    "df[\"UniProtID\"] = gene_ids_df.map(uniprot_ids)\n",
    "df[\"HGNC\"] = gene_ids_df.map(hgnc_symbols)\n",
    "\n",
    "# Save the updated data (replace with your desired output file path)\n",
    "df.to_excel(\"dengue_with_uniprot.xlsx\", index=False)\n",
    "df.head()  # Display the first few rows to verify the results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5d15833d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RNAseq 24hpi':     gene_symbol    log2FC  GeneID Condition\n",
       " 182         NaN -5.887024   45338       24h\n",
       " 183         NaN -2.734256   45338       24h\n",
       " 184         NaN  4.086980   45338       24h\n",
       " 185         NaN  2.314898   45338       24h,\n",
       " 'RNAseq 48hpi':     gene_symbol    log2FC  GeneID Condition\n",
       " 515         NaN  5.202878   45338       48h\n",
       " 516         NaN  5.169160   45338       48h\n",
       " 517         NaN  3.074556   45338       48h\n",
       " 518         NaN  2.817277   45338       48h\n",
       " 519         NaN  2.188067   45338       48h}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def debug_conflicts(data_dict):\n",
    "    # Define a list of dataframes to process\n",
    "    df_names = ['siRNA data', 'Pt Abundance', 'RNAseq 24hpi', 'RNAseq 48hpi']\n",
    "\n",
    "    # Define gene name columns for each dataframe\n",
    "    gene_name_cols = ['GeneSymbol', None, 'GeneName', 'GeneName']\n",
    "\n",
    "    # Define keys for each dataframe\n",
    "    keys = ['GeneID', 'GeneID', 'GeneID', 'GeneID']\n",
    "    \n",
    "    # Create a dictionary to hold conflict information\n",
    "    conflict_dict = {}\n",
    "\n",
    "    for df_name, key, gene_name in zip(df_names, keys, gene_name_cols):\n",
    "        df = data_dict[df_name].copy()\n",
    "\n",
    "        # check if gene name column exists in the dataframe\n",
    "        if gene_name and gene_name in df.columns:\n",
    "            df.rename(columns={gene_name: 'gene_symbol'}, inplace=True)\n",
    "            \n",
    "            # check if there are any conflicts in gene names for the same GeneID\n",
    "            conflicting_rows = df[df.duplicated(subset=[key, 'gene_symbol'], keep=False)]\n",
    "            if not conflicting_rows.empty:\n",
    "                conflict_dict[df_name] = conflicting_rows\n",
    "\n",
    "    return conflict_dict\n",
    "\n",
    "conflict_dict = debug_conflicts(data)\n",
    "\n",
    "conflict_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b10c3c",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "March 23\n",
    "Laura: There are not (yet) any approved drugs specific for dengue virus. Ribavirin is used as a treatment, but this is a generic antiviral and is not dengue specific. The promising small molecule JNJ-A07 is currently under investigation but it targets the virus directly and not the host cell. For host-directed targets that result in inhibition of dengue virus replication, some of them are listed in the article (please see Table 2): https://www.sciencedirect.com/science/article/pii/S1879625720300523  Hope this helps!\n",
    " \n",
    "Here is some keyword advice from ChatGPT:\n",
    " \n",
    "Here are some suggested keywords that you can search for on Google and PubMed if you are researching mechanisms of dengue infection:\n",
    "Dengue fever\n",
    "Dengue virus\n",
    "Viral replication\n",
    "Pathogenesis\n",
    "Host-pathogen interaction\n",
    "Immune response\n",
    "Dengue infection and cytokines\n",
    "Mosquito-borne diseases\n",
    "Viral entry\n",
    "Endocytosis\n",
    "Viral receptors\n",
    "Innate immune response\n",
    "Adaptive immune response\n",
    "Virus evasion strategies\n",
    "Dengue vaccine development\n",
    "You can also use specific gene or protein names that are known to be involved in dengue infection, such as NS1, NS2A, NS2B, NS3, NS4A, NS4B, and NS5. Additionally, you can try combining these keywords with terms such as \"molecular mechanisms\", \"pathophysiology\", \"epidemiology\", \"clinical features\", and \"diagnosis\", depending on your research focus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce553a04",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
