{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6432e006",
   "metadata": {},
   "source": [
    "## Goal Proteins\n",
    "\n",
    " - AND\n",
    "     - negatively affect dengue virus replication when inhibited by siRNA*\n",
    "     - OR\n",
    "        - are modulated in abundance\n",
    "        - are modulated in rna abundance\n",
    "        - interact with the dengue virus\n",
    "\n",
    "* \\*the highest the Zscore, the more these factors negatively affect dengue virus replication\n",
    "\n",
    "## Plan\n",
    " - Define rules implementing goal criteria\n",
    " - Separate the interaction data from the protein properties data\n",
    " - Combine the data sources to get a single protein table:\n",
    " - Make an network of interactions between the proteins: STRING HC + dengue AP-MS\n",
    " - Annotate the network with the protein data, properties including:\n",
    "     - Whether the host protein binds a viral protein\n",
    "     - A list of viral proteins that bind the protein\n",
    "     - whether it meets any of the interesting criteria: a protein of interest\n",
    "     - whether it is a goal protein\n",
    " - Make a hierarchical from the subnetwork interconnecting the proteins of interest\n",
    " - Name the model systems by enrichment\n",
    " - Name and some or all of the model systems by LLM\n",
    " - Identify goal proteins\n",
    " - Identify systems enriched with goal proteins\n",
    " - Style the model and the interactome to highlight the data\n",
    " - Generate reports\n",
    " - Generate figures\n",
    " \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe2451b",
   "metadata": {},
   "source": [
    "## Experimental Protocol\n",
    "We took primary human dendritic cells, infected them with dengue virus (serotype 3) and subjected these cells to either: \n",
    " - siRNA screening to identify human host factors that act to restrict viral replication, \n",
    " - Proteomics (Protein Abundance) to look at human proteins that change in abundance following infection. \n",
    "   - This was done at 24h and 48h post infection. (Jeff Johnson, Krogan lab)\n",
    " - RNAseq to look at cellular mRNAs that are differently expressed following infection.\n",
    "   - This was done at 24h and 48h post infection. (Stephen Wolinski, NWU)\n",
    "   \n",
    "We also include host-dengue PPI data from (Shah et al., Cell 2018). \n",
    "Priya Shah was at the Krogan lab when this data was collected\n",
    "\n",
    "The datasets are already thresholded based maximum overlap and significance, and all have GeneID as common identifier. \n",
    "\n",
    "20220202_Thresholded_Datasets_DHIPC.xlsx\n",
    "\n",
    "Shah et al., Dengue Interactions\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8caee487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "24936df3",
   "metadata": {},
   "source": [
    "## Load Data\n",
    "\n",
    "\n",
    "prompt: write a function load_dengue_data to load an excel spreadsheet as a dict of dataframes corresponding to the tabs in the spreadsheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0c9d3ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'siRNA data':      GeneName  GeneID  Average_Zscore\n",
       " 0       A2LD1   87769        1.588549\n",
       " 1       ABCC2    1244        1.209934\n",
       " 2       ABCG2    9429        1.293657\n",
       " 3       ACACB      32        1.126954\n",
       " 4      ACAS2L   84532        1.994937\n",
       " ...       ...     ...             ...\n",
       " 1058   ZNF704  619279        1.356330\n",
       " 1059   ZNF708    7562        1.527764\n",
       " 1060    ZNRF3   84133        1.363028\n",
       " 1061   ZSIG11   51368        1.436758\n",
       " 1062   ZWILCH   55055        1.406935\n",
       " \n",
       " [1063 rows x 3 columns],\n",
       " 'Pt Abundance':     Uniprot ID         Timepoint    log2FC  GeneID\n",
       " 0       Q86TX2  DV3_24h-Mock_24h  1.608293  641371\n",
       " 1       Q86YS6  DV3_24h-Mock_24h  1.346686  339122\n",
       " 2       Q8NFC6  DV3_48h-Mock_48h -1.871359  259282\n",
       " 3       Q96FN4  DV3_24h-Mock_24h  1.548070  221184\n",
       " 4       Q8IVG5  DV3_24h-Mock_24h  1.172951  219285\n",
       " ..         ...               ...       ...     ...\n",
       " 114     P16671  DV3_48h-Mock_48h -1.209983     948\n",
       " 115     P50454  DV3_48h-Mock_48h  1.071615     871\n",
       " 116     P30043  DV3_24h-Mock_24h  1.492406     645\n",
       " 117     P35573  DV3_24h-Mock_24h -1.062767     178\n",
       " 118     P01023  DV3_48h-Mock_48h -2.371772       2\n",
       " \n",
       " [119 rows x 4 columns],\n",
       " 'RNAseq 24hpi':        GeneName    log2FC     GeneID Condition\n",
       " 0        HS3ST2  2.299718       9956       24h\n",
       " 1        TRANK1  2.133498       9881       24h\n",
       " 2        ZBTB39 -2.249348       9880       24h\n",
       " 3       GPRASP1 -2.175409       9737       24h\n",
       " 4       FAM131B -2.200291       9715       24h\n",
       " ..          ...       ...        ...       ...\n",
       " 321     RAB1AP1  3.167020  100420571       24h\n",
       " 322  RNF213-AS1  4.731135  100294362       24h\n",
       " 323     LY6E-DT  3.012394  100133669       24h\n",
       " 324  SEMA3F-AS1 -2.167095  100129060       24h\n",
       " 325      MED6P1  6.554080  100128990       24h\n",
       " \n",
       " [326 rows x 4 columns],\n",
       " 'RNAseq 48hpi':     GeneName    log2FC  GeneID Condition\n",
       " 0      ABCA2 -2.435633      20       48h\n",
       " 1      ABCB9 -2.021558   23457       48h\n",
       " 2     ABLIM3 -4.255646   22885       48h\n",
       " 3      ABTB2  2.789291   25841       48h\n",
       " 4      ACOD1  6.432192  730249       48h\n",
       " ..       ...       ...     ...       ...\n",
       " 880   ZNF618  2.580672  114991       48h\n",
       " 881   ZNF623 -6.412347    9831       48h\n",
       " 882  ZNF702P  3.298258   79986       48h\n",
       " 883   ZNF711 -2.128946    7552       48h\n",
       " 884   ZNF724 -2.465182  440519       48h\n",
       " \n",
       " [885 rows x 4 columns],\n",
       " 'PPI data':                     Bait      MiST Uniprot ID  Gene ID Gene Symbol\n",
       " 0     DENV2 16681 Capsid  0.907662     Q96KR1    51663         ZFR\n",
       " 1     DENV2 16681 Capsid  0.903071     Q96DV4    64978      MRPL38\n",
       " 2     DENV2 16681 Capsid  0.896905     Q96SI9    55342       STRBP\n",
       " 3     DENV2 16681 Capsid  0.896263     Q9H9J2    65080      MRPL44\n",
       " 4     DENV2 16681 Capsid  0.874178     Q16540     6150      MRPL23\n",
       " ...                  ...       ...        ...      ...         ...\n",
       " 1825          ZIKVug NS5  0.616254     Q9BUQ8     9416       DDX23\n",
       " 1826          ZIKVug NS5  0.613256     P60228     3646       EIF3E\n",
       " 1827          ZIKVug NS5  0.607727     Q13620     8450       CUL4B\n",
       " 1828          ZIKVug NS5  0.601817     O75821     8666       EIF3G\n",
       " 1829          ZIKVug NS5  0.601440     Q7RTV0    84844       PHF5A\n",
       " \n",
       " [1830 rows x 5 columns]}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def load_dengue_data(file_path):\n",
    "    # Create an ExcelFile object\n",
    "    xls = pd.ExcelFile(file_path)\n",
    "    \n",
    "    # Get the sheet names\n",
    "    sheet_names = xls.sheet_names\n",
    "    \n",
    "    # Initialize a dictionary to store the dataframes\n",
    "    dengue_data = {}\n",
    "\n",
    "    # Loop over the sheet names and read each into a dataframe\n",
    "    for sheet in sheet_names:\n",
    "        dengue_data[sheet] = pd.read_excel(xls, sheet_name=sheet)\n",
    "    \n",
    "    return dengue_data\n",
    "\n",
    "data = load_dengue_data('20220202_Thresholded_Datasets_DHIPC.xlsx')\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10664e96",
   "metadata": {},
   "source": [
    "## Merge the RNA Data\n",
    "\n",
    "Prompt: write a function merge_dengue_data to operate on the data dict to merge these data frames : \n",
    "'siRNA data' - columns:  GeneName  GeneID  Average_Zscore\n",
    "'RNAseq 24hpi' - columns:  GeneName    log2FC  GeneID Condition\n",
    "'RNAseq 48hpi' - columns:  GeneName    log2FC  GeneID Condition\n",
    "\n",
    "Merge the dataframes using GeneID as the key\n",
    "Make the column names unique by prepending abbreviations for the dataframe names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fee0abaf",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'gene_symbol'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/12/1t8825t10yz4_h3t94kg8kyw0000gp/T/ipykernel_16931/1036749502.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m \u001b[0mmerged_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmerge_dengue_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0mmerged_data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/12/1t8825t10yz4_h3t94kg8kyw0000gp/T/ipykernel_16931/1036749502.py\u001b[0m in \u001b[0;36mmerge_dengue_data\u001b[0;34m(data_dict)\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m             \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34mf\"{abbr}_{col}\"\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m             \u001b[0mmerged_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmerged_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'outer'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_symbol'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m             \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34mf\"{abbr}_{col}\"\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m    108\u001b[0m     \u001b[0mvalidate\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    109\u001b[0m ) -> DataFrame:\n\u001b[0;32m--> 110\u001b[0;31m     op = _MergeOperation(\n\u001b[0m\u001b[1;32m    111\u001b[0m         \u001b[0mleft\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m         \u001b[0mright\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, indicator, validate)\u001b[0m\n\u001b[1;32m    701\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright_join_keys\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    702\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin_names\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 703\u001b[0;31m         ) = self._get_merge_keys()\n\u001b[0m\u001b[1;32m    704\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    705\u001b[0m         \u001b[0;31m# validate the merge keys dtypes. We may need to coerce\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m_get_merge_keys\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1160\u001b[0m                         \u001b[0mrk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mHashable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1161\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mrk\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1162\u001b[0;31m                             \u001b[0mright_keys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mright\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label_or_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1163\u001b[0m                         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1164\u001b[0m                             \u001b[0;31m# work-around for merge_asof(right_index=True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_label_or_level_values\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1848\u001b[0m             )\n\u001b[1;32m   1849\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1850\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1851\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1852\u001b[0m         \u001b[0;31m# Check for duplicates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'gene_symbol'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def merge_dengue_data(data_dict):\n",
    "    # Define a list of dataframes to merge\n",
    "    df_names = ['siRNA data', 'Pt Abundance', 'RNAseq 24hpi', 'RNAseq 48hpi']\n",
    "    \n",
    "    # Define abbreviations for the dataframe names\n",
    "    abbreviations = ['siRNA', 'PtAb', 'RNA24', 'RNA48']\n",
    "    \n",
    "    # Define gene name columns for each dataframe\n",
    "    gene_name_cols = ['GeneSymbol', None, 'GeneName', 'GeneName']\n",
    "    \n",
    "    # Define keys for each dataframe\n",
    "    keys = ['GeneID', 'GeneID', 'GeneID', 'GeneID']\n",
    "    \n",
    "    # Start with the first dataframe\n",
    "    merged_data = data_dict[df_names[0]].copy()\n",
    "    \n",
    "    # Check if gene name column exists in the dataframe\n",
    "    if gene_name_cols[0] in merged_data.columns:\n",
    "        merged_data.rename(columns={gene_name_cols[0]: 'gene_symbol'}, inplace=True)\n",
    "\n",
    "    merged_data.columns = [f\"{abbreviations[0]}_{col}\" if col != keys[0] else col for col in merged_data.columns]\n",
    "    \n",
    "    # Merge the remaining dataframes\n",
    "    for df_name, abbr, key, gene_name in zip(df_names[1:], abbreviations[1:], keys[1:], gene_name_cols[1:]):\n",
    "        df = data_dict[df_name].copy()\n",
    "\n",
    "        # check if gene name column exists in the dataframe\n",
    "        if gene_name and gene_name in df.columns:\n",
    "            df.rename(columns={gene_name: 'gene_symbol'}, inplace=True)\n",
    "\n",
    "            # drop the rows where gene_symbol is NaN before merging\n",
    "            df = df[df['gene_symbol'].notna()]\n",
    "            \n",
    "            df.columns = [f\"{abbr}_{col}\" if col != key else col for col in df.columns]\n",
    "            merged_data = pd.merge(merged_data, df, how='outer', on=[key, 'gene_symbol'])\n",
    "        else:\n",
    "            df.columns = [f\"{abbr}_{col}\" if col != key else col for col in df.columns]\n",
    "            merged_data = pd.merge(merged_data, df, how='outer', on=key)\n",
    "\n",
    "    return merged_data\n",
    "\n",
    "\n",
    "\n",
    "merged_data = merge_dengue_data(data)\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5d15833d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'RNAseq 24hpi':     gene_symbol    log2FC  GeneID Condition\n",
       " 182         NaN -5.887024   45338       24h\n",
       " 183         NaN -2.734256   45338       24h\n",
       " 184         NaN  4.086980   45338       24h\n",
       " 185         NaN  2.314898   45338       24h,\n",
       " 'RNAseq 48hpi':     gene_symbol    log2FC  GeneID Condition\n",
       " 515         NaN  5.202878   45338       48h\n",
       " 516         NaN  5.169160   45338       48h\n",
       " 517         NaN  3.074556   45338       48h\n",
       " 518         NaN  2.817277   45338       48h\n",
       " 519         NaN  2.188067   45338       48h}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def debug_conflicts(data_dict):\n",
    "    # Define a list of dataframes to process\n",
    "    df_names = ['siRNA data', 'Pt Abundance', 'RNAseq 24hpi', 'RNAseq 48hpi']\n",
    "\n",
    "    # Define gene name columns for each dataframe\n",
    "    gene_name_cols = ['GeneSymbol', None, 'GeneName', 'GeneName']\n",
    "\n",
    "    # Define keys for each dataframe\n",
    "    keys = ['GeneID', 'GeneID', 'GeneID', 'GeneID']\n",
    "    \n",
    "    # Create a dictionary to hold conflict information\n",
    "    conflict_dict = {}\n",
    "\n",
    "    for df_name, key, gene_name in zip(df_names, keys, gene_name_cols):\n",
    "        df = data_dict[df_name].copy()\n",
    "\n",
    "        # check if gene name column exists in the dataframe\n",
    "        if gene_name and gene_name in df.columns:\n",
    "            df.rename(columns={gene_name: 'gene_symbol'}, inplace=True)\n",
    "            \n",
    "            # check if there are any conflicts in gene names for the same GeneID\n",
    "            conflicting_rows = df[df.duplicated(subset=[key, 'gene_symbol'], keep=False)]\n",
    "            if not conflicting_rows.empty:\n",
    "                conflict_dict[df_name] = conflicting_rows\n",
    "\n",
    "    return conflict_dict\n",
    "\n",
    "conflict_dict = debug_conflicts(data)\n",
    "\n",
    "conflict_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b10c3c",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "March 23\n",
    "Laura: There are not (yet) any approved drugs specific for dengue virus. Ribavirin is used as a treatment, but this is a generic antiviral and is not dengue specific. The promising small molecule JNJ-A07 is currently under investigation but it targets the virus directly and not the host cell. For host-directed targets that result in inhibition of dengue virus replication, some of them are listed in the article (please see Table 2): https://www.sciencedirect.com/science/article/pii/S1879625720300523  Hope this helps!\n",
    " \n",
    "Here is some keyword advice from ChatGPT:\n",
    " \n",
    "Here are some suggested keywords that you can search for on Google and PubMed if you are researching mechanisms of dengue infection:\n",
    "Dengue fever\n",
    "Dengue virus\n",
    "Viral replication\n",
    "Pathogenesis\n",
    "Host-pathogen interaction\n",
    "Immune response\n",
    "Dengue infection and cytokines\n",
    "Mosquito-borne diseases\n",
    "Viral entry\n",
    "Endocytosis\n",
    "Viral receptors\n",
    "Innate immune response\n",
    "Adaptive immune response\n",
    "Virus evasion strategies\n",
    "Dengue vaccine development\n",
    "You can also use specific gene or protein names that are known to be involved in dengue infection, such as NS1, NS2A, NS2B, NS3, NS4A, NS4B, and NS5. Additionally, you can try combining these keywords with terms such as \"molecular mechanisms\", \"pathophysiology\", \"epidemiology\", \"clinical features\", and \"diagnosis\", depending on your research focus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce553a04",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
